# cointab challlange


You are a data analyst and have a large ecommerce company in India (let’s call it X) as a client.
X gets a few thousand orders via their website on a daily basis and they have to deliver them as
fast as they can. For delivering the goods ordered by the customers, X has tied up with multiple
courier companies in India who charge them some amount per delivery.


The charges are dependent upon two factors:

● Weight of the product.

● Distance between the warehouse (pickup location) and customer’s delivery address
(destination location)


On an average, the delivery charges are Rs. 100 per shipment. So if X ships 1,00,000 orders
per month, they have to pay approximately Rs. 1 crore to the courier companies on a monthly
basis as charges.


As the amount that X has to pay to the courier companies is very high, they want to verify if the
charges levied by their partners per Order are correct.

Importing pandas and numpy libraries for reading csv files and operations on csv files.


In [1]:
import numpy as np
import pandas as pd

In [2]:
order_report_df=pd.read_excel('Company X - Order Report.xlsx')

In [3]:
len(order_report_df["ExternOrderNo"].unique())

124

In [4]:
len(order_report_df['SKU'].unique())

65

In [5]:
order_report_df

ExternOrderNo            SKU  Order Qty
0       2001827036  8904223818706        1.0
1       2001827036  8904223819093        1.0
2       2001827036  8904223819109        1.0
3       2001827036  8904223818430        1.0
4       2001827036  8904223819277        1.0
..             ...            ...        ...
395     2001806229  8904223818942        1.0
396     2001806229  8904223818850        1.0
397     2001806226  8904223818850        2.0
398     2001806210  8904223816214        1.0
399     2001806210  8904223818874        1.0

[400 rows x 3 columns]

In [6]:
order_report_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ExternOrderNo  400 non-null    int64  
 1   SKU            400 non-null    object 
 2   Order Qty      400 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 9.5+ KB


finding duplicate rows

In [7]:
order_report_df.duplicated().any()

True

In [8]:
print("duplicate rows in order_report_df")
order_report_df[order_report_df.duplicated()]

duplicate rows in order_report_df


ExternOrderNo            SKU  Order Qty
110     2001811363  8904223815859        1.0
317     2001806885  8904223819499        2.0

In [9]:
order_report_df[order_report_df.duplicated()==False]

ExternOrderNo            SKU  Order Qty
0       2001827036  8904223818706        1.0
1       2001827036  8904223819093        1.0
2       2001827036  8904223819109        1.0
3       2001827036  8904223818430        1.0
4       2001827036  8904223819277        1.0
..             ...            ...        ...
395     2001806229  8904223818942        1.0
396     2001806229  8904223818850        1.0
397     2001806226  8904223818850        2.0
398     2001806210  8904223816214        1.0
399     2001806210  8904223818874        1.0

[398 rows x 3 columns]

SKU(stock keeping unit), in full stock keeping unit, a code number,
typically used as a machine-readable bar code, assigned to a single item of inventory

In [10]:
pincode_zones_df=pd.read_excel('Company X - Pincode Zones.xlsx')

In [11]:
order_pincodes_df=pincode_zones_df[pincode_zones_df.duplicated()==False]
order_pincodes_df

Warehouse Pincode  Customer Pincode Zone
0               121003            507101    d
1               121003            486886    d
2               121003            532484    d
3               121003            143001    b
4               121003            515591    d
..                 ...               ...  ...
114             121003            324008    b
115             121003            302020    b
119             121003            325207    b
120             121003            303702    b
121             121003            313301    b

[108 rows x 3 columns]

In [12]:
len(pincode_zones_df["Customer Pincode"].unique())

108

In [13]:
pincode_zones_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Warehouse Pincode  124 non-null    int64 
 1   Customer Pincode   124 non-null    int64 
 2   Zone               124 non-null    object
dtypes: int64(2), object(1)
memory usage: 3.0+ KB


In [14]:
sku_master_df=pd.read_excel('Company X - SKU Master.xlsx')

In [15]:
sku_master_df

SKU  Weight (g)
0   8904223815682         210
1   8904223815859         165
2   8904223815866         113
3   8904223815873          65
4   8904223816214         120
..            ...         ...
61  8904223819505         210
62  8904223819499         210
63  8904223819512         210
64  8904223819543         300
65     SACHETS001          10

[66 rows x 2 columns]

In [16]:
sku_master_df["Weight (g)"].max()

600

In [17]:
sku_master_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   SKU         66 non-null     object
 1   Weight (g)  66 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.2+ KB



The number of distinct SKU's in order info and sku master table are same. 

we have summarise the company order info into 124 rows from 400 rows. 

For this, we have to check for null values and duplicate rows and agggregate the rows.

In [18]:
len(sku_master_df["SKU"].unique())

65

In [19]:
sku_master_df[sku_master_df.duplicated()]

SKU  Weight (g)
56  GIFTBOX202002         500

In [20]:
sku_master_df[sku_master_df.duplicated()==False]

SKU  Weight (g)
0   8904223815682         210
1   8904223815859         165
2   8904223815866         113
3   8904223815873          65
4   8904223816214         120
..            ...         ...
61  8904223819505         210
62  8904223819499         210
63  8904223819512         210
64  8904223819543         300
65     SACHETS001          10

[65 rows x 2 columns]

In [21]:
X_order_df=pd.merge(order_report_df[order_report_df.duplicated()==False],sku_master_df[sku_master_df.duplicated()==False],on='SKU')

In [22]:
X_order_df

ExternOrderNo            SKU  Order Qty  Weight (g)
0       2001827036  8904223818706        1.0         127
1       2001821995  8904223818706        1.0         127
2       2001819252  8904223818706        1.0         127
3       2001816996  8904223818706        1.0         127
4       2001814580  8904223818706        1.0         127
..             ...            ...        ...         ...
393     2001806616  8904223819123        1.0         250
394     2001806567  8904223815804        1.0         160
395     2001806567  8904223818577        1.0         150
396     2001806408  8904223819437        2.0         552
397     2001806232  8904223818645        6.0         137

[398 rows x 4 columns]

In [23]:
X_order_df["weight(kg)"]=X_order_df["Order Qty"]*X_order_df["Weight (g)"]/1000

In [24]:
X_order_df

ExternOrderNo            SKU  Order Qty  Weight (g)  weight(kg)
0       2001827036  8904223818706        1.0         127       0.127
1       2001821995  8904223818706        1.0         127       0.127
2       2001819252  8904223818706        1.0         127       0.127
3       2001816996  8904223818706        1.0         127       0.127
4       2001814580  8904223818706        1.0         127       0.127
..             ...            ...        ...         ...         ...
393     2001806616  8904223819123        1.0         250       0.250
394     2001806567  8904223815804        1.0         160       0.160
395     2001806567  8904223818577        1.0         150       0.150
396     2001806408  8904223819437        2.0         552       1.104
397     2001806232  8904223818645        6.0         137       0.822

[398 rows x 5 columns]

In [25]:
X_order_df.groupby(["ExternOrderNo"])["weight(kg)"].sum()

ExternOrderNo
2001806210    0.220
2001806226    0.480
2001806229    0.500
2001806232    1.302
2001806233    0.245
              ...  
2001821995    0.477
2001822466    1.376
2001823564    0.672
2001825261    1.557
2001827036    1.676
Name: weight(kg), Length: 124, dtype: float64

In [26]:
order_df=pd.DataFrame(X_order_df.groupby(["ExternOrderNo"])["weight(kg)"].sum())

In [27]:
order_df

weight(kg)
ExternOrderNo            
2001806210          0.220
2001806226          0.480
2001806229          0.500
2001806232          1.302
2001806233          0.245
...                   ...
2001821995          0.477
2001822466          1.376
2001823564          0.672
2001825261          1.557
2001827036          1.676

[124 rows x 1 columns]

In [28]:
order_df["Order ID"]=order_df.index

In [29]:
order_df

weight(kg)    Order ID
ExternOrderNo                        
2001806210          0.220  2001806210
2001806226          0.480  2001806226
2001806229          0.500  2001806229
2001806232          1.302  2001806232
2001806233          0.245  2001806233
...                   ...         ...
2001821995          0.477  2001821995
2001822466          1.376  2001822466
2001823564          0.672  2001823564
2001825261          1.557  2001825261
2001827036          1.676  2001827036

[124 rows x 2 columns]

In [30]:
courier_invoice_df=pd.read_excel('Courier Company - Invoice.xlsx')

In [31]:
courier_invoice_df

AWB Code    Order ID  Charged Weight  Warehouse Pincode  \
0    1091117222124  2001806232            1.30             121003   
1    1091117222194  2001806273            1.00             121003   
2    1091117222931  2001806408            2.50             121003   
3    1091117223244  2001806458            1.00             121003   
4    1091117229345  2001807012            0.15             121003   
..             ...         ...             ...                ...   
119  1091118551656  2001812941            0.73             121003   
120  1091117614452  2001809383            0.50             121003   
121  1091120922803  2001820978            0.50             121003   
122  1091121844806  2001811475            0.50             121003   
123  1091121846136  2001811305            0.50             121003   

     Customer Pincode Zone         Type of Shipment  Billing Amount (Rs.)  
0              507101    d          Forward charges                 135.0  
1              486886    d          Forward charges                  90.2  
2              532484    d          Forward charges                 224.6  
3              143001    b          Forward charges                  61.3  
4              515591    d          Forward charges                  45.4  
..                ...  ...                      ...                   ...  
119            325207    d          Forward charges                  90.2  
120            303702    d  Forward and RTO charges                  86.7  
121            313301    d          Forward charges                  45.4  
122            173212    b          Forward charges                  33.0  
123            302020    d          Forward charges                  45.4  

[124 rows x 8 columns]

In [32]:
courier_invoice_df["Charged Weight"].max()

4.13

In [33]:
courier_rates=pd.read_excel('Courier Company - Rates.xlsx')
courier_rates

fwd_a_fixed  fwd_a_additional  fwd_b_fixed  fwd_b_additional  fwd_c_fixed  \
0         29.5              23.6           33              28.3         40.1   

   fwd_c_additional  fwd_d_fixed  fwd_d_additional  fwd_e_fixed  \
0              38.9         45.4              44.8         56.6   

   fwd_e_additional  rto_a_fixed  rto_a_additional  rto_b_fixed  \
0              55.5         13.6              23.6         20.5   

   rto_b_additional  rto_c_fixed  rto_c_additional  rto_d_fixed  \
0              28.3         31.9              38.9         41.3   

   rto_d_additional  rto_e_fixed  rto_e_additional  
0              44.8         50.7              55.5

In [34]:
courier_invoice_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   AWB Code              124 non-null    int64  
 1   Order ID              124 non-null    int64  
 2   Charged Weight        124 non-null    float64
 3   Warehouse Pincode     124 non-null    int64  
 4   Customer Pincode      124 non-null    int64  
 5   Zone                  124 non-null    object 
 6   Type of Shipment      124 non-null    object 
 7   Billing Amount (Rs.)  124 non-null    float64
dtypes: float64(2), int64(4), object(2)
memory usage: 7.9+ KB


In [35]:
len(courier_invoice_df["Order ID"].unique())

124

In [36]:
len(courier_invoice_df["AWB Code"].unique())

124

In [37]:
result_1=pd.merge(order_df,courier_invoice_df,on='Order ID')
result_1

weight(kg)    Order ID       AWB Code  Charged Weight  Warehouse Pincode  \
0         0.220  2001806210  1091117221940            2.92             121003   
1         0.480  2001806226  1091117222065            0.68             121003   
2         0.500  2001806229  1091117222080            0.71             121003   
3         1.302  2001806232  1091117222124            1.30             121003   
4         0.245  2001806233  1091117222135            0.78             121003   
..          ...         ...            ...             ...                ...   
119       0.477  2001821995  1091121183730            0.50             121003   
120       1.376  2001822466  1091121305541            1.10             121003   
121       0.672  2001823564  1091121666133            0.70             121003   
122       1.557  2001825261  1091121981575            1.60             121003   
123       1.676  2001827036  1091122418320            1.60             121003   

     Customer Pincode Zone         Type of Shipment  Billing Amount (Rs.)  
0              140604    b          Forward charges                 174.5  
1              723146    d          Forward charges                  90.2  
2              421204    d          Forward charges                  90.2  
3              507101    d          Forward charges                 135.0  
4              263139    b          Forward charges                  61.3  
..                ...  ...                      ...                   ...  
119            342008    d          Forward charges                  45.4  
120            342301    d          Forward charges                 135.0  
121            492001    d  Forward and RTO charges                 172.8  
122            517128    d  Forward and RTO charges                 345.0  
123            173213    b          Forward charges                 117.9  

[124 rows x 9 columns]

In [38]:
dataframe_1=pd.merge(order_pincodes_df,result_1,on=['Warehouse Pincode','Customer Pincode'])
dataframe_1

Warehouse Pincode  Customer Pincode Zone_x  weight(kg)    Order ID  \
0               121003            507101      d       1.302  2001806232   
1               121003            486886      d       0.615  2001806273   
2               121003            532484      d       2.265  2001806408   
3               121003            143001      b       0.700  2001806458   
4               121003            515591      d       0.240  2001807012   
..                 ...               ...    ...         ...         ...   
119             121003            302020      b       0.488  2001811039   
120             121003            302020      b       0.750  2001811305   
121             121003            325207      b       0.500  2001812941   
122             121003            303702      b       0.607  2001809383   
123             121003            313301      b       0.515  2001820978   

          AWB Code  Charged Weight Zone_y         Type of Shipment  \
0    1091117222124            1.30      d          Forward charges   
1    1091117222194            1.00      d          Forward charges   
2    1091117222931            2.50      d          Forward charges   
3    1091117223244            1.00      b          Forward charges   
4    1091117229345            0.15      d          Forward charges   
..             ...             ...    ...                      ...   
119  1091117904860            0.68      d          Forward charges   
120  1091121846136            0.50      d          Forward charges   
121  1091118551656            0.73      d          Forward charges   
122  1091117614452            0.50      d  Forward and RTO charges   
123  1091120922803            0.50      d          Forward charges   

     Billing Amount (Rs.)  
0                   135.0  
1                    90.2  
2                   224.6  
3                    61.3  
4                    45.4  
..                    ...  
119                  90.2  
120                  45.4  
121                  90.2  
122                  86.7  
123                  45.4  

[124 rows x 10 columns]

In [39]:
dataframe_1.rename(columns={'Zone_x':'Delivery Zone as per X', 'weight(kg)':'Total weight as per X (KG)',
                            'Charged Weight':'Total weight as per Courier Company (KG)',
                            'Zone_y':'Delivery Zone charged by Courier Company',
                           'Billing Amount (Rs.)':'Charges Billed by Courier Company (Rs.)'}, inplace = True)

In [40]:
dataframe_1

Warehouse Pincode  Customer Pincode Delivery Zone as per X  \
0               121003            507101                      d   
1               121003            486886                      d   
2               121003            532484                      d   
3               121003            143001                      b   
4               121003            515591                      d   
..                 ...               ...                    ...   
119             121003            302020                      b   
120             121003            302020                      b   
121             121003            325207                      b   
122             121003            303702                      b   
123             121003            313301                      b   

     Total weight as per X (KG)    Order ID       AWB Code  \
0                         1.302  2001806232  1091117222124   
1                         0.615  2001806273  1091117222194   
2                         2.265  2001806408  1091117222931   
3                         0.700  2001806458  1091117223244   
4                         0.240  2001807012  1091117229345   
..                          ...         ...            ...   
119                       0.488  2001811039  1091117904860   
120                       0.750  2001811305  1091121846136   
121                       0.500  2001812941  1091118551656   
122                       0.607  2001809383  1091117614452   
123                       0.515  2001820978  1091120922803   

     Total weight as per Courier Company (KG)  \
0                                        1.30   
1                                        1.00   
2                                        2.50   
3                                        1.00   
4                                        0.15   
..                                        ...   
119                                      0.68   
120                                      0.50   
121                                      0.73   
122                                      0.50   
123                                      0.50   

    Delivery Zone charged by Courier Company         Type of Shipment  \
0                                          d          Forward charges   
1                                          d          Forward charges   
2                                          d          Forward charges   
3                                          b          Forward charges   
4                                          d          Forward charges   
..                                       ...                      ...   
119                                        d          Forward charges   
120                                        d          Forward charges   
121                                        d          Forward charges   
122                                        d  Forward and RTO charges   
123                                        d          Forward charges   

     Charges Billed by Courier Company (Rs.)  
0                                      135.0  
1                                       90.2  
2                                      224.6  
3                                       61.3  
4                                       45.4  
..                                       ...  
119                                     90.2  
120                                     45.4  
121                                     90.2  
122                                     86.7  
123                                     45.4  

[124 rows x 10 columns]

In [41]:
dataframe_1["Type of Shipment"].unique()

array(['Forward charges', 'Forward and RTO charges'], dtype=object)

In [42]:
dataframe_1.drop(['Warehouse Pincode', 'Customer Pincode'], axis = 1,inplace=True)

In [43]:
dataframe_1

Delivery Zone as per X  Total weight as per X (KG)    Order ID  \
0                        d                       1.302  2001806232   
1                        d                       0.615  2001806273   
2                        d                       2.265  2001806408   
3                        b                       0.700  2001806458   
4                        d                       0.240  2001807012   
..                     ...                         ...         ...   
119                      b                       0.488  2001811039   
120                      b                       0.750  2001811305   
121                      b                       0.500  2001812941   
122                      b                       0.607  2001809383   
123                      b                       0.515  2001820978   

          AWB Code  Total weight as per Courier Company (KG)  \
0    1091117222124                                      1.30   
1    1091117222194                                      1.00   
2    1091117222931                                      2.50   
3    1091117223244                                      1.00   
4    1091117229345                                      0.15   
..             ...                                       ...   
119  1091117904860                                      0.68   
120  1091121846136                                      0.50   
121  1091118551656                                      0.73   
122  1091117614452                                      0.50   
123  1091120922803                                      0.50   

    Delivery Zone charged by Courier Company         Type of Shipment  \
0                                          d          Forward charges   
1                                          d          Forward charges   
2                                          d          Forward charges   
3                                          b          Forward charges   
4                                          d          Forward charges   
..                                       ...                      ...   
119                                        d          Forward charges   
120                                        d          Forward charges   
121                                        d          Forward charges   
122                                        d  Forward and RTO charges   
123                                        d          Forward charges   

     Charges Billed by Courier Company (Rs.)  
0                                      135.0  
1                                       90.2  
2                                      224.6  
3                                       61.3  
4                                       45.4  
..                                       ...  
119                                     90.2  
120                                     45.4  
121                                     90.2  
122                                     86.7  
123                                     45.4  

[124 rows x 8 columns]

In [44]:
# importing math library for using ceiling function
# this can not be use on arrays directly
# only works on numbers
import math

In [45]:
math.ceil(0.3)

1

In [46]:
dataframe_1['Total weight as per X (KG)'].apply(np.ceil)*0.5

0      1.0
1      0.5
2      1.5
3      0.5
4      0.5
      ... 
119    0.5
120    0.5
121    0.5
122    0.5
123    0.5
Name: Total weight as per X (KG), Length: 124, dtype: float64

In [47]:
dataframe_1['Weight slab as per X (KG)']=(dataframe_1['Total weight as per X (KG)']/0.5).apply(np.ceil)*0.5

In [48]:
dataframe_1['Weight slab charged by Courier Company (KG)']= (dataframe_1['Total weight as per Courier Company (KG)']/0.5).apply(np.ceil)*0.5

In [49]:
dataframe_1

Delivery Zone as per X  Total weight as per X (KG)    Order ID  \
0                        d                       1.302  2001806232   
1                        d                       0.615  2001806273   
2                        d                       2.265  2001806408   
3                        b                       0.700  2001806458   
4                        d                       0.240  2001807012   
..                     ...                         ...         ...   
119                      b                       0.488  2001811039   
120                      b                       0.750  2001811305   
121                      b                       0.500  2001812941   
122                      b                       0.607  2001809383   
123                      b                       0.515  2001820978   

          AWB Code  Total weight as per Courier Company (KG)  \
0    1091117222124                                      1.30   
1    1091117222194                                      1.00   
2    1091117222931                                      2.50   
3    1091117223244                                      1.00   
4    1091117229345                                      0.15   
..             ...                                       ...   
119  1091117904860                                      0.68   
120  1091121846136                                      0.50   
121  1091118551656                                      0.73   
122  1091117614452                                      0.50   
123  1091120922803                                      0.50   

    Delivery Zone charged by Courier Company         Type of Shipment  \
0                                          d          Forward charges   
1                                          d          Forward charges   
2                                          d          Forward charges   
3                                          b          Forward charges   
4                                          d          Forward charges   
..                                       ...                      ...   
119                                        d          Forward charges   
120                                        d          Forward charges   
121                                        d          Forward charges   
122                                        d  Forward and RTO charges   
123                                        d          Forward charges   

     Charges Billed by Courier Company (Rs.)  Weight slab as per X (KG)  \
0                                      135.0                        1.5   
1                                       90.2                        1.0   
2                                      224.6                        2.5   
3                                       61.3                        1.0   
4                                       45.4                        0.5   
..                                       ...                        ...   
119                                     90.2                        0.5   
120                                     45.4                        1.0   
121                                     90.2                        0.5   
122                                     86.7                        1.0   
123                                     45.4                        1.0   

     Weight slab charged by Courier Company (KG)  
0                                            1.5  
1                                            1.0  
2                                            2.5  
3                                            1.0  
4                                            0.5  
..                                           ...  
119                                          1.0  
120                                          0.5  
121                                          1.0  
122                                          0.5  
123                                          0.5  

[12

In [50]:
dataframe_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 124 entries, 0 to 123
Data columns (total 10 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   Delivery Zone as per X                       124 non-null    object 
 1   Total weight as per X (KG)                   124 non-null    float64
 2   Order ID                                     124 non-null    int64  
 3   AWB Code                                     124 non-null    int64  
 4   Total weight as per Courier Company (KG)     124 non-null    float64
 5   Delivery Zone charged by Courier Company     124 non-null    object 
 6   Type of Shipment                             124 non-null    object 
 7   Charges Billed by Courier Company (Rs.)      124 non-null    float64
 8   Weight slab as per X (KG)                    124 non-null    float64
 9   Weight slab charged by Courier Company (KG)  124 non-null    float64
dtypes:

● Courier charges rate card at weight slab and pincode level. If the invoice mentions
“Forward charges” then only forward charges (“fwd”) should be applicable as per zone
and fixed & additional weights based on weight slabs. If the invoice mentions “Forward
and rto charges” then forward charges (“fwd”) and RTO charges (“rto”) should be
applicable as per zone and fixed & additional weights based on weight slabs.


● For the first 0.5 KG, “fixed” rate as per the slab is applicable. For each additional 0.5 KG,
“additional” weight in the same proportion is applicable. Total charges will be “fixed” +
“total additional” if any


In [51]:
courier_rates

fwd_a_fixed  fwd_a_additional  fwd_b_fixed  fwd_b_additional  fwd_c_fixed  \
0         29.5              23.6           33              28.3         40.1   

   fwd_c_additional  fwd_d_fixed  fwd_d_additional  fwd_e_fixed  \
0              38.9         45.4              44.8         56.6   

   fwd_e_additional  rto_a_fixed  rto_a_additional  rto_b_fixed  \
0              55.5         13.6              23.6         20.5   

   rto_b_additional  rto_c_fixed  rto_c_additional  rto_d_fixed  \
0              28.3         31.9              38.9         41.3   

   rto_d_additional  rto_e_fixed  rto_e_additional  
0              44.8         50.7              55.5

In [52]:
dataframe_1["Type of Shipment"].value_counts()

Forward charges            109
Forward and RTO charges     15
Name: Type of Shipment, dtype: int64

In [53]:
dataframe_1["Delivery Zone as per X"].value_counts()

b    79
d    38
e     7
Name: Delivery Zone as per X, dtype: int64

In [54]:
dataframe_1["Delivery Zone charged by Courier Company"].value_counts()

d    98
b    24
e     2
Name: Delivery Zone charged by Courier Company, dtype: int64

● From above, we can say we have only two types of couriers. They are

● Forward charges      --- >     109

● Forward and RTO charges  --->   15

● To get expected charges as per X company, we have to use some conditional statements.



In [55]:
# we have "b","d","e" zones in company order info

def calculated_cost(dataframe_1):
    if (dataframe_1["Type of Shipment"]=="Forward charges"):
        if (dataframe_1["Delivery Zone as per X"]=='a'):
            
            if (dataframe_1["Weight slab as per X (KG)"]<=0.5):
                return courier_rates["fwd_a_fixed"]
            else:
                return ((dataframe_1["Weight slab as per X (KG)"]-0.5)*2*courier_rates["fwd_a_additional"])+courier_rates["fwd_a_fixed"]
            
        elif (dataframe_1["Delivery Zone as per X"]=='b'):
            
            if (dataframe_1["Weight slab as per X (KG)"]<=0.5):
                return courier_rates["fwd_b_fixed"]
            else:
                return ((dataframe_1["Weight slab as per X (KG)"]-0.5)*2*courier_rates["fwd_b_additional"])+courier_rates["fwd_b_fixed"]
        
            
             
        elif (dataframe_1["Delivery Zone as per X"]=='d'):
            
            if (dataframe_1["Weight slab as per X (KG)"]<=0.5):
                return courier_rates["fwd_d_fixed"]
            else:
                return ((dataframe_1["Weight slab as per X (KG)"]-0.5)*2*courier_rates["fwd_d_additional"])+courier_rates["fwd_d_fixed"]   
        
        
        elif (dataframe_1["Delivery Zone as per X"]=='e'):
            
            if (dataframe_1["Weight slab as per X (KG)"]<=0.5):
                return courier_rates["fwd_e_fixed"]
            else:
                return ((dataframe_1["Weight slab as per X (KG)"]-0.5)*2*courier_rates["fwd_e_additional"])+courier_rates["fwd_e_fixed"]   
            
    else:
        if (dataframe_1["Delivery Zone as per X"]=='b'):
            
            if (dataframe_1["Weight slab as per X (KG)"]<=0.5):
                return courier_rates["fwd_b_fixed"]+courier_rates["rto_b_fixed"]
            else:
                return ((dataframe_1["Weight slab as per X (KG)"]-0.5)*2*(courier_rates["fwd_b_additional"]+courier_rates["rto_b_additional"]))+courier_rates["fwd_b_fixed"]+courier_rates["rto_b_fixed"]
        
        
        elif (dataframe_1["Delivery Zone as per X"]=='d'):
            if (dataframe_1["Weight slab as per X (KG)"]<=0.5):
                return courier_rates["fwd_d_fixed"]+courier_rates["rto_d_fixed"]
            else:
                return ((dataframe_1["Weight slab as per X (KG)"]-0.5)*2*(courier_rates["fwd_d_additional"]+courier_rates["rto_d_additional"]))+courier_rates["fwd_d_fixed"]+courier_rates["rto_d_fixed"]

            
        elif (dataframe_1["Delivery Zone as per X"]=='e'):
             if (dataframe_1["Weight slab as per X (KG)"]<=0.5):
                return courier_rates["fwd_e_fixed"]+courier_rates["rto_e_fixed"]
             else:
                return ((dataframe_1["Weight slab as per X (KG)"]-0.5)*2*(courier_rates["fwd_e_additional"]+courier_rates["rto_e_additional"]))+courier_rates["fwd_e_fixed"]+courier_rates["rto_e_fixed"]

            
            
        

In [56]:
dataframe_1['Expected Charge as per X (Rs.)'] = dataframe_1.apply(calculated_cost, axis = 1)

In [57]:
dataframe_1

Delivery Zone as per X  Total weight as per X (KG)    Order ID  \
0                        d                       1.302  2001806232   
1                        d                       0.615  2001806273   
2                        d                       2.265  2001806408   
3                        b                       0.700  2001806458   
4                        d                       0.240  2001807012   
..                     ...                         ...         ...   
119                      b                       0.488  2001811039   
120                      b                       0.750  2001811305   
121                      b                       0.500  2001812941   
122                      b                       0.607  2001809383   
123                      b                       0.515  2001820978   

          AWB Code  Total weight as per Courier Company (KG)  \
0    1091117222124                                      1.30   
1    1091117222194                                      1.00   
2    1091117222931                                      2.50   
3    1091117223244                                      1.00   
4    1091117229345                                      0.15   
..             ...                                       ...   
119  1091117904860                                      0.68   
120  1091121846136                                      0.50   
121  1091118551656                                      0.73   
122  1091117614452                                      0.50   
123  1091120922803                                      0.50   

    Delivery Zone charged by Courier Company         Type of Shipment  \
0                                          d          Forward charges   
1                                          d          Forward charges   
2                                          d          Forward charges   
3                                          b          Forward charges   
4                                          d          Forward charges   
..                                       ...                      ...   
119                                        d          Forward charges   
120                                        d          Forward charges   
121                                        d          Forward charges   
122                                        d  Forward and RTO charges   
123                                        d          Forward charges   

     Charges Billed by Courier Company (Rs.)  Weight slab as per X (KG)  \
0                                      135.0                        1.5   
1                                       90.2                        1.0   
2                                      224.6                        2.5   
3                                       61.3                        1.0   
4                                       45.4                        0.5   
..                                       ...                        ...   
119                                     90.2                        0.5   
120                                     45.4                        1.0   
121                                     90.2                        0.5   
122                                     86.7                        1.0   
123                                     45.4                        1.0   

     Weight slab charged by Courier Company (KG)  \
0                                            1.5   
1                                            1.0   
2                                            2.5   
3                                            1.0   
4                                            0.5   
..                                           ...   
119                                          1.0   
120                                          0.5   
121                                          1.0   
122                                          0.5   
123                                         

In [58]:
dataframe_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 124 entries, 0 to 123
Data columns (total 11 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   Delivery Zone as per X                       124 non-null    object 
 1   Total weight as per X (KG)                   124 non-null    float64
 2   Order ID                                     124 non-null    int64  
 3   AWB Code                                     124 non-null    int64  
 4   Total weight as per Courier Company (KG)     124 non-null    float64
 5   Delivery Zone charged by Courier Company     124 non-null    object 
 6   Type of Shipment                             124 non-null    object 
 7   Charges Billed by Courier Company (Rs.)      124 non-null    float64
 8   Weight slab as per X (KG)                    124 non-null    float64
 9   Weight slab charged by Courier Company (KG)  124 non-null    float64
 10  Ex

In [59]:
dataframe_1["Difference Between Expected Charges and Billed Charges (Rs.)"]=dataframe_1["Expected Charge as per X (Rs.)"]-dataframe_1["Charges Billed by Courier Company (Rs.)"]

In [60]:
dataframe_1

Delivery Zone as per X  Total weight as per X (KG)    Order ID  \
0                        d                       1.302  2001806232   
1                        d                       0.615  2001806273   
2                        d                       2.265  2001806408   
3                        b                       0.700  2001806458   
4                        d                       0.240  2001807012   
..                     ...                         ...         ...   
119                      b                       0.488  2001811039   
120                      b                       0.750  2001811305   
121                      b                       0.500  2001812941   
122                      b                       0.607  2001809383   
123                      b                       0.515  2001820978   

          AWB Code  Total weight as per Courier Company (KG)  \
0    1091117222124                                      1.30   
1    1091117222194                                      1.00   
2    1091117222931                                      2.50   
3    1091117223244                                      1.00   
4    1091117229345                                      0.15   
..             ...                                       ...   
119  1091117904860                                      0.68   
120  1091121846136                                      0.50   
121  1091118551656                                      0.73   
122  1091117614452                                      0.50   
123  1091120922803                                      0.50   

    Delivery Zone charged by Courier Company         Type of Shipment  \
0                                          d          Forward charges   
1                                          d          Forward charges   
2                                          d          Forward charges   
3                                          b          Forward charges   
4                                          d          Forward charges   
..                                       ...                      ...   
119                                        d          Forward charges   
120                                        d          Forward charges   
121                                        d          Forward charges   
122                                        d  Forward and RTO charges   
123                                        d          Forward charges   

     Charges Billed by Courier Company (Rs.)  Weight slab as per X (KG)  \
0                                      135.0                        1.5   
1                                       90.2                        1.0   
2                                      224.6                        2.5   
3                                       61.3                        1.0   
4                                       45.4                        0.5   
..                                       ...                        ...   
119                                     90.2                        0.5   
120                                     45.4                        1.0   
121                                     90.2                        0.5   
122                                     86.7                        1.0   
123                                     45.4                        1.0   

     Weight slab charged by Courier Company (KG)  \
0                                            1.5   
1                                            1.0   
2                                            2.5   
3                                            1.0   
4                                            0.5   
..                                           ...   
119                                          1.0   
120                                          0.5   
121                                          1.0   
122                                          0.5   
123                                         

In [61]:
dataframe_1.drop(['Type of Shipment'], axis = 1,inplace=True)

In [62]:
dataframe_1

Delivery Zone as per X  Total weight as per X (KG)    Order ID  \
0                        d                       1.302  2001806232   
1                        d                       0.615  2001806273   
2                        d                       2.265  2001806408   
3                        b                       0.700  2001806458   
4                        d                       0.240  2001807012   
..                     ...                         ...         ...   
119                      b                       0.488  2001811039   
120                      b                       0.750  2001811305   
121                      b                       0.500  2001812941   
122                      b                       0.607  2001809383   
123                      b                       0.515  2001820978   

          AWB Code  Total weight as per Courier Company (KG)  \
0    1091117222124                                      1.30   
1    1091117222194                                      1.00   
2    1091117222931                                      2.50   
3    1091117223244                                      1.00   
4    1091117229345                                      0.15   
..             ...                                       ...   
119  1091117904860                                      0.68   
120  1091121846136                                      0.50   
121  1091118551656                                      0.73   
122  1091117614452                                      0.50   
123  1091120922803                                      0.50   

    Delivery Zone charged by Courier Company  \
0                                          d   
1                                          d   
2                                          d   
3                                          b   
4                                          d   
..                                       ...   
119                                        d   
120                                        d   
121                                        d   
122                                        d   
123                                        d   

     Charges Billed by Courier Company (Rs.)  Weight slab as per X (KG)  \
0                                      135.0                        1.5   
1                                       90.2                        1.0   
2                                      224.6                        2.5   
3                                       61.3                        1.0   
4                                       45.4                        0.5   
..                                       ...                        ...   
119                                     90.2                        0.5   
120                                     45.4                        1.0   
121                                     90.2                        0.5   
122                                     86.7                        1.0   
123                                     45.4                        1.0   

     Weight slab charged by Courier Company (KG)  \
0                                            1.5   
1                                            1.0   
2                                            2.5   
3                                            1.0   
4                                            0.5   
..                                           ...   
119                                          1.0   
120                                          0.5   
121                                          1.0   
122                                          0.5   
123                                          0.5   

     Expected Charge as per X (Rs.)  \
0                             135.0   
1                              90.2   
2                             224.6   
3                              61.3   
4                              45.4   
..                              ...   
119               

In [63]:
# writing to Excel
datatoexcel = pd.ExcelWriter('result_1.xlsx')
  
# write DataFrame to excel
dataframe_1.to_excel(datatoexcel)
  
# save the excel
datatoexcel.save()
print('DataFrame is written to Excel File successfully.')

DataFrame is written to Excel File successfully.


In [64]:
dataframe_1["Difference Between Expected Charges and Billed Charges (Rs.)"].sum()

-3908.1

In [65]:
dataframe_1["Charges Billed by Courier Company (Rs.)"].sum()

13648.2

In [66]:
dataframe_1["Expected Charge as per X (Rs.)"].sum()

9740.099999999997

In [69]:
dataframe_1[dataframe_1["Difference Between Expected Charges and Billed Charges (Rs.)"]==0]

Delivery Zone as per X  Total weight as per X (KG)    Order ID  \
0                       d                       1.302  2001806232   
2                       d                       2.265  2001806408   
3                       b                       0.700  2001806458   
4                       d                       0.240  2001807012   
5                       d                       0.240  2001806686   
7                       b                       1.168  2001807058   
8                       b                       0.343  2001808118   
9                       d                       0.500  2001807186   
10                      d                       0.500  2001807290   
11                      b                       0.607  2001807814   
13                      b                       1.080  2001807956   
15                      d                       0.240  2001807930   
16                      d                       1.157  2001808102   
19                      b                       0.607  2001808507   
21                      b                       1.183  2001808675   

         AWB Code  Total weight as per Courier Company (KG)  \
0   1091117222124                                      1.30   
2   1091117222931                                      2.50   
3   1091117223244                                      1.00   
4   1091117229345                                      0.15   
5   1091117229555                                      0.15   
7   1091117323112                                      1.15   
8   1091117435134                                      0.50   
9   1091117323812                                      0.50   
10  1091117324206                                      0.50   
11  1091117326612                                      0.79   
13  1091117327275                                      1.08   
15  1091117327695                                      0.15   
16  1091117435005                                      1.28   
19  1091117436383                                      0.79   
21  1091117437050                                      1.20   

   Delivery Zone charged by Courier Company  \
0                                         d   
2                                         d   
3                                         b   
4                                         d   
5                                         d   
7                                         b   
8                                         b   
9                                         d   
10                                        d   
11                                        b   
13                                        b   
15                                        d   
16                                        d   
19                                        b   
21                                        b   

    Charges Billed by Courier Company (Rs.)  Weight slab as per X (KG)  \
0                                     135.0                        1.5   
2                                     224.6                        2.5   
3                                      61.3                        1.0   
4                                      45.4                        0.5   
5                                      45.4                        0.5   
7                                      89.6                        1.5   
8                                      33.0                        0.5   
9                                      45.4                        0.5   
10                                     45.4                        0.5   
11                                     61.3                        1.0   
13                                     89.6                        1.5   
15                                     45.4                        0.5   
16                                    135.0                        1.5   
19                                     61.3                        1.0   
21                        

In [75]:
dataframe_1[dataframe_1["Difference Between Expected Charges and Billed Charges (Rs.)"]==0].shape

(15, 11)

In [71]:
dataframe_1[dataframe_1["Difference Between Expected Charges and Billed Charges (Rs.)"]==0]['Charges Billed by Courier Company (Rs.)'].sum()

1207.3

above vale is Total orders where X has been correctly charged and their count is 15 and bill amount is 1207.3 RS

In [79]:
dataframe_1[dataframe_1["Difference Between Expected Charges and Billed Charges (Rs.)"]<0].shape

(85, 11)

In [80]:
dataframe_1[dataframe_1["Difference Between Expected Charges and Billed Charges (Rs.)"]<0]['Charges Billed by Courier Company (Rs.)'].sum()

8982.100000000004

above reslt is Total Orders where X has been overcharged. the count is 85 and total bill is 8982.1 rs

In [81]:
dataframe_1[dataframe_1["Difference Between Expected Charges and Billed Charges (Rs.)"]>0].shape

(24, 11)

In [82]:
dataframe_1[dataframe_1["Difference Between Expected Charges and Billed Charges (Rs.)"]>0]['Charges Billed by Courier Company (Rs.)'].sum()

3458.8

above reslt is Total Orders where X has been under charged. the count is 24 and total bill is 3458.8 rs